File: backend/app/core/security.py
"""Security helpers: password hashing and JWT tokens.

Password hashing uses PBKDF2-HMAC-SHA256 from the Python standard library
(OWASP-approved, no external C dependencies). Tokens are signed HS256 JWTs.
"""
import hashlib
import hmac
import secrets
from datetime import datetime, timedelta, timezone
from typing import Optional

import jwt

from app.core.config import settings

_PBKDF2_ITERATIONS = 390_000
_ALGORITHM = "HS256"


def hash_password(password: str) -> str:
    salt = secrets.token_hex(16)
    digest = hashlib.pbkdf2_hmac(
        "sha256", password.encode("utf-8"), bytes.fromhex(salt), _PBKDF2_ITERATIONS
    )
    return f"pbkdf2_sha256${_PBKDF2_ITERATIONS}${salt}${digest.hex()}"


def verify_password(password: str, stored: str) -> bool:
    try:
        _, iterations, salt, expected = stored.split("$")
        digest = hashlib.pbkdf2_hmac(
            "sha256", password.encode("utf-8"), bytes.fromhex(salt), int(iterations)
        )
        return hmac.compare_digest(digest.hex(), expected)
    except (ValueError, TypeError):
        return False


def create_access_token(user_id: int, expires_minutes: Optional[int] = None) -> str:
    minutes = expires_minutes or settings.ACCESS_TOKEN_EXPIRE_MINUTES
    now = datetime.now(timezone.utc)
    payload = {
        "sub": str(user_id),
        "iat": now,
        "exp": now + timedelta(minutes=minutes),
        "type": "access",
    }
    return jwt.encode(payload, settings.SECRET_KEY, algorithm=_ALGORITHM)


def decode_access_token(token: str) -> Optional[int]:
    """Return the user id if the token is valid, otherwise None."""
    try:
        payload = jwt.decode(token, settings.SECRET_KEY, algorithms=[_ALGORITHM])
        if payload.get("type") != "access":
            return None
        return int(payload["sub"])
    except (jwt.PyJWTError, ValueError, KeyError):
        return None


def random_secret(length: int = 48) -> str:
    return secrets.token_urlsafe(length)

File: backend/src/security.ts
import crypto from "node:crypto";
import jwt from "jsonwebtoken";
import { config } from "./config.js";

const PBKDF2_ITERATIONS = 390_000;
const KEY_LENGTH = 64;

export function hashPassword(password: string): string {
  const salt = crypto.randomBytes(16).toString("hex");
  const digest = crypto.pbkdf2Sync(password, salt, PBKDF2_ITERATIONS, KEY_LENGTH, "sha256").toString("hex");
  return `pbkdf2_sha256$${PBKDF2_ITERATIONS}$${salt}$${digest}`;
}

export function verifyPassword(password: string, stored: string): boolean {
  try {
    const [, iterations, salt, expected] = stored.split("$");
    const digest = crypto
      .pbkdf2Sync(password, salt, Number(iterations), KEY_LENGTH, "sha256")
      .toString("hex");
    return crypto.timingSafeEqual(Buffer.from(digest), Buffer.from(expected));
  } catch {
    return false;
  }
}

export interface TokenPayload {
  sub: string;
  type: "access";
}

export function createAccessToken(userId: string): string {
  return jwt.sign(
    {
      sub: userId,
      type: "access",
    },
    config.jwtSecret,
    {
      algorithm: "HS256",
      expiresIn: `${config.accessTokenExpireMinutes}m`,
    }
  );
}

export function decodeAccessToken(token: string): string | null {
  try {
    const payload = jwt.verify(token, config.jwtSecret, { algorithms: ["HS256"] }) as TokenPayload;
    if (payload.type !== "access" || !payload.sub) return null;
    return payload.sub;
  } catch {
    return null;
  }
}
